lib/rules/semi.js

```diff
@@ -166,6 +168,30 @@ module.exports = {
             );
         }
 
+        /**
+         * Checks if a given PropertyDefinition node followed by a semicolon
+         * can safely remove that semicolon. It is not to safe to remove if
+         * the class field name is "get", "set", or "static", or if
+         * followed by a generator method.
+         * @param {ASTNode} node The node to check.
+         * @returns {boolean} `true` if the node cannot have the semicolon
+         *      removed.
+         */
+        function maybeClassFieldAsiHazard(node) {
+
+            if (node.type !== "PropertyDefinition") {
+                return false;
+            }
+
+            if (unsafeClassFieldNames.has(node.key.name)) {
+                return true;
+            }
+
+            const followingToken = sourceCode.getTokenAfter(node);
+
+            return unsafeClassFieldFollowers.has(followingToken.value);
```

Comment:
Because these aren't statements, it seems like `beforeStatementContinuationChars` shouldn't apply here. I'll fix that.

Reply:
Okay, I thought I understood what you are saying here, but I don't. I don't think `beforeStatementContinuationChars` should apply inside of classes, but I'm unclear on what the expected behavior would be in that case. Could you provide an example of what is happening now with the current state of this PR and what you're expecting if `beforeStatementContinuationChars` doesn't apply inside of classes?

Reply:
Expected behavior:

```js
/* eslint semi: ["error", "never", { beforeStatementContinuationChars: "never" }] */

class C {
    x // no linting errors here
    [foo]

    x; // no linting errors here
    [foo]

    x = "a"; // no linting errors here
    [foo]
}
```

Actual behavior:

```js
/* eslint semi: ["error", "never", { beforeStatementContinuationChars: "never" }] */

class C {
    x // no linting errors here
    [foo]

    x; // linting error (1)
    [foo]

    x = "a"; // linting error (2)
    [foo]
}
```

(1) The expected behavior that this isn't a linting error, even though this `;` can be safely removed, is based on the default behavior for statements with the `"never"` option, which is that the rule allows both `;` and no `;` before tokens that can continue statements before them (e.g., `[`), despite the fact that the particular statement before the token cannot be continued (e.g., `ImportDeclaration`).

```js
/* eslint semi: ["error", "never"] */

import a from "a" // ok
[foo]

import b from "b"; // also ok
[foo]
```

So if we ever add an option for class elements that would be equivalent to `beforeStatementContinuationChars`, its default will be the same as the default for `beforeStatementContinuationChars`, which is `"any"`.

(2) This is a bug regardless of what we decide about the options, because removing this `;` turns two class field definitions `x = "a"` and `[foo]` into one  `x = "a"[foo]`.

Reply:
Expected behavior with `beforeStatementContinuationChars: "always"` (same as in the previous comment with `"never"`):

```js
/* eslint semi: ["error", "never", { beforeStatementContinuationChars: "always" }] */

class C {
    x // no linting errors here
    [foo]

    x; // no linting errors here
    [foo]

    x = 1; // no linting errors here
    [foo]
}
```

Actual behavior:

```js
/* eslint semi: ["error", "never", { beforeStatementContinuationChars: "always" }] */

class C {
    x // linting error
    [foo]

    x; // no linting errors here
    [foo]

    x = 1; // no linting errors here
    [foo]
}
```



The `beforeStatementContinuationChars` option definitely affects class elements in the current state of this PR, since changing its value produces different linting results in the same code.